realm/rfc/0001-backlink-query-syntax.md

```diff
@@ -0,0 +1,204 @@
+realm.where(Dog.class).equalTo("name[Person.favoriteDog]").findAll();
+
+// Example in a purely text based variant
+"name[Person] = 'John'"
+```
+
+**Implementation**
+Implementation is straight forward as we have all schema information available 
+when parsing queries. So simple String manipulation will make it possible to 
+split a field name into a name and a type part.
+
+Validating that the type is an actual backlink with the given field does require 
+an `O(n)` lookup where `m: number of fields in Person`. Since number of fields 
+are generally low, this shouldn't be a problem in practise.
+
+**Advantages**
+- Using "[]" is not an allowed field characters in any of our supported 
+  languages as far as I know. So parsing the above expressions should be 
+  possible with no ambiguity.
+
+- The default case is easy to describe.
+
+- Using [] as a modifier already has precedent in NSPredicate.
+
+
+**Disadvantages**
+
+- Unknown syntax. This concept is not used nor supported in neither NSPredicate, 
+  LINQ or SQL. It will be a "power feature", and might make the language feel 
+  complicated and hard to use.
+
+- Moves the Java query language further away from being type safe.
+
+- Will introduce two ways of querying backlinks.
+
+
+
+## Alternative syntax suggestions
+
+As an alternative to the above, I have listed a some of variants that were 
+considered:
+
+```
+1) Type in front
+
+- "[Person]name = 'John'"
+- "[Person.favoriteDog]name = 'John'"
+
+
+2) All fields explicit. Group fields using ()
+
+- "name[Person.(favoriteDog, dogs)]  = 'John'"
+
+
+3) Other modifiers (^ up) , (<- back)
+
+- "^Person(favoriteDog, dogs).name"
+- "Person(favoriteDog, dogs)<-.name"
+
+```
+
+## Possible extensions
+
+The proposed query syntax can also be extended to support the following features:
+
+```
+// Wild cards
+- "name[*] = 'John' " // All backlink types
+- "name[* extends Person] = 'John' " // Polymorphic backlinks
+- *[*] = 'John' // Any string based field on all types of backlinks
+
+However this is not part of this RFC.
+
+```
+
+
+# Drawbacks
+
+It is not clear that his approach is the correct way forward. I am not sure what 
+can be done to gain more confidence in this, expect asking for more input. Please 
+see the list of advantages and disadvantages in the last section as well as the 
+proposed alternatives.
+
+
+# Alternatives
+
+A number of alternatives exists:
+
+## 1. Only named backlinks
+
+Objective C / Swift only allow named backlinks that are defined as part
+of the Realm model class.
+
+**Advantages:**
+- Minimal effort required to implement backlink queries. Current query system is 
+  enough.
+- Already supported by Objective C / Swift
+
+**Disadvantages:**
+- Does not work _at all_ in the dynamic API.
+- User ambiguity regarding manual migrations: Some fields must be added in a 
+  migration, others not needed. This is already happening with ignored fields
+  though.
```

Comment:
No, it should be possible to do async queries on them as well. owners is not a TableView but a BacklinkView (I won't claim I fully understand the difference yet though). I am just re-using RealmResults instead of inventing a new class to hold them. Cocoa had to implement an entire new class due to other constraints. I am hoping we can avoid that since, for the end user, it could semantically be seen as a view into underlying data (RealmResults).